classes/submission/Collector.inc.php

```diff
@@ -46,13 +52,40 @@ public function filterBySectionIds(array $sectionIds): self
         return $this;
     }
 
+    /**
+     * Limit results to submissions that contain any pub objects (e.g. publication and galley) with these statuses
+     *
+     * @param array|null $statuses One or more of DOI::STATUS_* constants
+     * @param bool $strict Whether null/empty values should count when considering Doi::STATUS_UNREGISTERED
+     *
+     * @return $this
+     */
+    public function filterByAnyDoiStatuses(?array $statuses, bool $strict = false): self
```

Comment:
That sounds good. If I move the filter to pkp-lib, and the actual query is app specific, is it problematic to have the actual query in the app-specific `getQueryBuilder` and have the variables and `filterByDoiStatus` in pkp-lib?

Reply:
What do you mean by app-specific? Don't DOIs have a status in all applications? And aren't these statuses the same?

Reply:
Yeah, that's right. This was for filtering submissions by the DOI status of any of their sub pub objects, so for OJS this would mean publications and galleys and OMP: publications, chapters, publication formats, and submission files. The app-specific part referred to the [SQL query to filter by this](https://github.com/ewhanson/ojs/blob/08e13de46e0b8ce799427aa568a603c84a9b6520/classes/submission/Collector.inc.php#L88-L107).

Because the pub objects could have DOI objects with different statuses (e.g. a publication that is registered and a galley that is unregistered), when filtering by unregistered, we would want the submission containing that galley to show up.

My thinking was to move the function/property declaration (`filterByDoiStatus()`, `$publicationDoiStatuses`, etc) to `pkp-lib` and keep the OJS/OMP/OPS-specific query that's a part of `getQueryBuilder` in the APP scope.

This would look something like:

pkp-lib Submission Collector
```PHP
...
public ?array $doiStatuses = null;
/** @var bool Whether null/empty values should count when considering Doi::STATUS_UNREGISTERED */
public bool $strictDoiStatusFilter = false;
...
/**
 * Limit results to submissions that contain any pub objects (e.g. publication and galley) with these statuses
 *
 * @param array|null $statuses One or more of DOI::STATUS_* constants
 * @param bool $strict Whether null/empty values should count when considering Doi::STATUS_UNREGISTERED
 *
 * @return AppCollector
 */
public function filterByDoiStatuses(?array $statuses, bool $strict = false): AppCollector
{
    $this->doiStatuses = $statuses;
    $this->strictDoiStatusFilter = $strict;
    return $this;
}
...
```

OJS Submission Collector
```PHP
...
/**
 * @copydoc CollectorInterface::getQueryBuilder()
 */
public function getQueryBuilder(): Builder
{
    ...
            // By any child pub object's DOI status
            $q->when($this->doiStatuses !== null, function (Builder  $q) {
                $q->whereIn('s.current_publication_id', function (Builder $q) {
                    $q->select('current_p.publication_id')
                        ->from('publications as current_p')
                        ->leftJoin('publication_galleys as current_g', 'current_g.publication_id', '=', 'current_p.publication_id')
                        ->leftJoin('dois as pd', 'pd.doi_id', '=', 'current_p.doi_id')
                        ->leftJoin('dois as gd', 'gd.doi_id', '=', 'current_g.doi_id')
                        ->whereIn('pd.status', $this->doiStatuses)
                        ->orWhereIn('gd.status', $this->doiStatuses);

                $q->when(
                    (in_array(Doi::STATUS_UNREGISTERED, $this->doiStatuses) && !$this->strictDoiStatusFilter),
                    function (Builder $q) {
                        $q->orWhereNull('pd.status')
                            ->orWhereNull('gd.status');
                    }
                );
            });
        });
    ...
}
...
```

Reply:
Right! I get it now. Yes, that's a good way to do it. One thing you can do to enforce the support is use an abstract method in the pkp-lib collector. This will then throw a fatal error if the app-specific collector forgets to implement it:

```php
namespace PKP\...;

abstract class Collector
{
    public function getQueryBuilder(): Builder
    {
        ....
        $this->addDOIStatusFilterToQuery($q);
        ....
    }

    abstract protected function addDOIStatusFilterToQuery($q)
    {
        $q->when(...);
    }
}
```

That makes `addDOIStatusFilterToQuery` like a "contract" that the parent class makes with any child class, so that it can force a child class to implement it.